### PR TITLE
Autotools build system

### LINKED ISSUE #99: automake build system

### PR DESCRIPTION
**Pull request by wizzard**
- Implemented autotools based build system (closes #99)
- Fixed several compatibility issues
- Extended and renamed retrace wrapper script
- Extended documentation

### CODE REVIEW

**Comment by wizzard** [comment]
please run `make check` to perform tests. 
I'm currently debugging the issue with hanging test (it hangs at cmockatest.c:804).

**Comment by wizzard** [comment]
@efiop could we please ignore generated files (such as Makefile and test/Makefile) from checkpatch scanning? 
Thank you!

File: retrace.in [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+        output="$path"
+        return 0
+    fi
+
+    return 1
+}
+
+
+main() {
+	[ $# -lt 1  ] && \
```
Comment (efiop) [inline retrace.in]: why does it have such crazy formatting? Tab is used here, but not down below. Starange line splitting, etc.

Comment (wizzard) [inline retrace.in]: Will fix the formatting, thanks for pointing.

**Comment by efiop** [comment]
@wizzard Thanks for the fix, but the check will probably still fail, as every single commit in the pull request is checked. Could you squash your fixes into corresponding commits please?

**Comment by wizzard** [comment]
Will do, but please review your tool, as squashing commits is not the right way to submit changes. Usually each commit is a single finished task, which can be then referred, tested, rolled back. And usually commits have messages, where developers describe what was done and why.

**Comment by efiop** [comment]
@wizzard I agree, but every commit should be a single meaningful finished piece that doesn't break anything. If you have commit1 and then N commits like "fix formating in commit1" this is not a right way to do it too, as it makes project history a mess. For such commits git has a --fixup option(or manual git rebase -i HEAD~N and squashing and fixing-up).

So, for example, if you have such commits in your pull-request:
commit1
commit2
fix for commit1
commit3
fix 2 for commit1
...
you should rebase them and squash(fixup) with meaningful commit msgs and so on, to make your pull-request look like:
commit1
commit2
commit3
so it is easy to review your pull request and follow it in the history when it is merged.

**Comment by wizzard** [comment]
Squashed my commits into a single commit.
Most non-autotools related changes are insignificant and do not require separate commits.

**Comment by wizzard** [comment]
@efiop Could we please merge this branch with the master tree?
The issue related with test failing on MacOS does not belong to this task / branch.

**Comment by efiop** [comment]
@wizzard It seems like the problems are directly related to your patch, as it has Mac detection and it definitely should not break the build. Most of the issues are "undeclared identifier" for calls that are not present on Mac. Please correct me if I'm wrong, but new build system should've automatically detected present calls and changed macros appropriately, right?

**Comment by wizzard** [comment]
@efiop absolutely right, the new system automatically detects and enables / disables system specific features.
The current build issue is not related to the new build system, as this test file previously was disabled on MacOs (see the master's tree test/Makefile). 
I'd like to merge this branch first, so other guys could start using the build system asap, and then I'll send another patch which fixes cmockatest.c.

**Comment by efiop** [comment]
@wizzard I see. But could you disable it for now too? If we merge this now, other guys will always get broken builds in travis for MacOS and thus something important might slip out of our attention.

**Comment by wizzard** [comment]
@efiop done, disabled cmocka tests on non Linux OS for now. Will send MacOs path tomorrow in a separate pull request.

**Comment by efiop** [comment]
@wizzard Thank you!

**Pull request merged**